daml-foundations/daml-ghc/tests/ContractKeyTests.daml

```diff
@@ -0,0 +1,84 @@
+-- @IGNORE-LF 1.2
+daml 1.2
+module ContractKeyTests where
+
+-- (1)  If `key` mentions `this.x` then `maintainer` can mention
+-- `this.x.y`.
+
+data Course =
+  Course with
+      institution : Party
+      title : Text
+  deriving (Show, Eq)
+
+data Registration =
+  Registration with
+      student : Party
+      course : Course
+      year : Int
+  deriving (Show, Eq)
+
+template Enrollment
+  with
+      reg : Registration
+  where
+      signatory reg.student, reg.course.institution
+      key reg : Registration -- this.x
+      maintainer reg.course.institution -- this.x.y.z
+
+-- (2)
+
+data M = M { m : Party } deriving (Show, Eq)
+data K = K { s : M } deriving (Show, Eq)
+
+template T
+  with
+    i : Party
+    k : K
+  where
+    signatory i
+    key k.s : M
+    maintainer this.k.s.m -- This doesn't work without qualifying with
+                          -- `this`. Is that expected?
```

Comment:
I am surprised. This should work without the `this`. I'll investigate. Let's merge this for now and I'll fix it in a separate PR.

Reply:
Ok perfect. Glad I raised it with you!